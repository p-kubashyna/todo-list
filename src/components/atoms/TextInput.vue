<script setup>
import { ref } from "vue";

const inputText = ref("");
const emit = defineEmits(["submitForm"]);
const props = defineProps({
  placeholder: {
    type: String,
    required: true,
    default: "",
  },
  isInvalidInput: {
    type: Boolean,
    required: false,
    default: false,
  },
});

function submitEvent() {
  emit("submitForm", inputText.value);
  inputText.value = "";
}
</script>

<template>
  <form @submit.prevent="submitEvent()">
    <input
      class="input-base"
      :class="{ 'input-invalid': props.isInvalidInput }"
      v-model.trim="inputText"
      :placeholder="props.placeholder"
    />
  </form>
</template>

<style scoped>
.input-base {
  width: 600px;
  height: 25px;
  border-radius: 7px;
  border-width: 10px;
  border-color: white;
  border-style: solid;
  font-family: monospace;
}

.input-invalid {
  background: white url(../../assets/images/invalid_input_icon.png) no-repeat
    right;
  background-size: auto 100%;
}
</style>
