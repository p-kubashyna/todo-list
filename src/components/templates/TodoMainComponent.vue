<script setup>
/* eslint-disable */
import TextInput from "@/components/atoms/TextInput.vue";
import TodoPanel from "@/components/organisms/TodoPanel.vue";
import TodoTitle from "@/components/atoms/TodoTitle.vue";
import BackgroundImage from "@/components/atoms/BackgroundImage.vue";
import { ref } from "vue";
import store from "@/store";

const isInvalidInput = ref(false);

function submitTodo(inputText) {
  isInvalidInput.value = validateInput(inputText);
  if (!isInvalidInput.value) {
    store.commit("addTodo", inputText);
  }
}

function validateInput(inputText) {
  return !inputText || /^\s*$/.test(inputText) || inputText.length > 150;
}
</script>

<template>
  <BackgroundImage />
  <div class="content">
    <TodoTitle />
    <div>
      <TextInput @submitForm="(inputText) => submitTodo(inputText)" :is-invalid-input=isInvalidInput placeholder="Add task..."/>
      <TodoPanel />
    </div>
  </div>
</template>

<style>
.content {
  position: relative;
  padding: 150px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: -webkit-center;
}
</style>
