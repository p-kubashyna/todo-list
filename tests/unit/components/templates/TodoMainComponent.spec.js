import { describe, expect, it } from "@jest/globals";
import { mount, shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";
import TodoMainComponent from "@/components/templates/TodoMainComponent.vue";
import TextInput from "@/components/atoms/TextInput.vue";

describe("Given the TodoMainComponent component", () => {
  it("should render", () => {
    const store = createStore({});
    const wrapper = shallowMount(TodoMainComponent, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should add new valid todo", async () => {
    const store = createStore({});
    store.commit = jest.fn();
    const wrapper = mount(TodoMainComponent, {
      global: {
        plugins: [store],
      },
    });
    await wrapper.findComponent(TextInput).vm.$emit("submitForm", "todo");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(1);
    expect(store.commit).toHaveBeenCalledWith("addTodo", "todo");
  });

  it.each(["  ", "", "X".repeat(151)])(
    "should not add invalid todo",
    async (invalidTodo) => {
      const store = createStore({});
      store.commit = jest.fn();
      const wrapper = mount(TodoMainComponent, {
        global: {
          plugins: [store],
        },
      });
      await wrapper
        .findComponent(TextInput)
        .vm.$emit("submitForm", invalidTodo);
      await wrapper.vm.$nextTick();

      expect(store.commit).toHaveBeenCalledTimes(0);
    }
  );
});
