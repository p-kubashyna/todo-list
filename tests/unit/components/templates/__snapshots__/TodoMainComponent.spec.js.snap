// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TodoMainComponent component should render 1`] = `
<div
  data-v-app=""
>
  
  <background-image-stub />
  <div
    class="content"
  >
    <todo-title-stub />
    <div>
      <text-input-stub
        isinvalidinput="false"
        placeholder="Add task..."
      />
      <todo-panel-stub />
    </div>
  </div>
  
</div>
`;
